### PR TITLE
drivers: usb: fix a buffer overflow in usb_sam0.c

### PR DESCRIPTION
**Pull request by nzmichaelh**
The driver currently blindly copies all of the outgoing bytes into the
endpoint.  Instead, calculate the endpoint size and copy up to that
amount instead.

Signed-off-by: Michael Hope <mlhx@google.com>

### CODE REVIEW

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: drivers/usb/device/usb_dc_sam0.c [hunk 1]
```diff
@@ -551,6 +552,16 @@ int usb_dc_ep_write(u8_t ep, const u8_t *buf, u32_t len, u32_t *ret_bytes)
 		return -EAGAIN;
 	}
 
+	/* Turn the encoded buffer size back into bytes */
+	if (mps >= 7) {
```
[the remark this answers is not in the document]
Comment (nzmichaelh) [reply drivers/usb/device/usb_dc_sam0.c]: Done.

The driver supports variable packet sizes.  ep_buf is used as an allocate only heap.

File: drivers/usb/device/usb_dc_sam0.c [hunk 2]
```diff
@@ -551,6 +552,16 @@ int usb_dc_ep_write(u8_t ep, const u8_t *buf, u32_t len, u32_t *ret_bytes)
...
+		mps = 1023;
+	} else {
+		mps = 1 << (mps+3);
```
[the remark this answers is not in the document]
Comment (nzmichaelh) [reply drivers/usb/device/usb_dc_sam0.c]: Done by refactoring.

File: drivers/usb/device/usb_dc_sam0.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jfischer-phytec-iot) [reply drivers/usb/device/usb_dc_sam0.c]: > The driver supports variable packet sizes. ep_buf is used as an allocate only heap.

yeah, but it is valid to disable the endpoint and free allocated memory. And there are not any checks for available memory, data->brk can always be incremented. Please fix it as well, please use Memory Management provided by Zephyr RTOS.

[the remark this answers is not in the document]
Comment (nzmichaelh) [reply drivers/usb/device/usb_dc_sam0.c]: I'm happy to do that, but it's existing code separate issue to this PR.  OK to submit this PR and change the memory allocation in a follow up?

[the remark this answers is not in the document]
Comment (jfischer-phytec-iot) [reply drivers/usb/device/usb_dc_sam0.c]: yes, thanks

**Review by jfischer-no — APPROVED** [review]

File: drivers/usb/device/usb_dc_sam0.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (nzmichaelh) [reply drivers/usb/device/usb_dc_sam0.c]: A WIP version using the heap is in https://github.com/nzmichaelh/zephyr/tree/use_heap.  It depends on this PR being merged first.

[the remark this answers is not in the document]
Comment (emob-nordic) [reply drivers/usb/device/usb_dc_sam0.c]: > Please add an enum type for the Endpoint Size, like
> 
> ```
>  enum usb_sam0_endpoint_size {
>     USB_SAM0_EPSIZE_8    = 0,
>     USB_SAM0_EPSIZE_16,
> ...
> }
> ```

Why like so? Is it gonna be used somewhere else in the future? If no why not simply:

``` diff
diff --git a/drivers/usb/device/usb_dc_sam0.c b/drivers/usb/device/usb_dc_sam0.c
index 1c14bd099e..f95ee33d52 100644
--- a/drivers/usb/device/usb_dc_sam0.c
+++ b/drivers/usb/device/usb_dc_sam0.c
@@ -27,27 +27,8 @@ LOG_MODULE_REGISTER(usb_dc_sam0);
 #define USB_NUM_ENDPOINTS DT_INST_PROP(0, num_bidir_endpoints)
 
 /* The endpoint size stored in USB.PCKSIZE.SIZE */
-enum usb_sam0_pcksize_size {
-	USB_SAM0_PCKSIZE_SIZE_8 = 0,
-	USB_SAM0_PCKSIZE_SIZE_16,
-	USB_SAM0_PCKSIZE_SIZE_32,
-	USB_SAM0_PCKSIZE_SIZE_64,
-	USB_SAM0_PCKSIZE_SIZE_128,
-	USB_SAM0_PCKSIZE_SIZE_256,
-	USB_SAM0_PCKSIZE_SIZE_512,
-	USB_SAM0_PCKSIZE_SIZE_1023,
-};
-
 static const uint16_t usb_sam0_pcksize_bytes[] = {
-	[USB_SAM0_PCKSIZE_SIZE_8] = 8,
-	[USB_SAM0_PCKSIZE_SIZE_16] = 16,
-	[USB_SAM0_PCKSIZE_SIZE_32] = 32,
-	[USB_SAM0_PCKSIZE_SIZE_64] = 64,
-	[USB_SAM0_PCKSIZE_SIZE_128] = 128,
-	[USB_SAM0_PCKSIZE_SIZE_256] = 256,
-	[USB_SAM0_PCKSIZE_SIZE_512] = 512,
-	[USB_SAM0_PCKSIZE_SIZE_1023] = 1023,
-};
+	8, 16, 32, 64, 128, 256, 512, 1023};
 
 BUILD_ASSERT(ARRAY_SIZE(usb_sam0_pcksize_bytes) == 8);
 
```

[the remark this answers is not in the document]
Comment (nzmichaelh) [reply drivers/usb/device/usb_dc_sam0.c]: Hi @emob-nordic I'm not really worried - all of these are equivalent, about the same flash size, and as maintainable.  Unless you have a strong reason for the array then I'd rather keep the enum based reason.

[the remark this answers is not in the document]
Comment (emob-nordic) [reply drivers/usb/device/usb_dc_sam0.c]: This is a bunch of unnecessary code which could be replaced one line, I don't see any reason to create this enum.

[the remark this answers is not in the document]
Comment (jfischer-phytec-iot) [reply drivers/usb/device/usb_dc_sam0.c]: I have requested for the enum, @nzmichaelh is the author of this driver and submitted IMHO better solution, so I am fine with the changes. @nzmichaelh It is your decision to keep the enum or remove it :-)

[the remark this answers is not in the document]
Comment (nzmichaelh) [reply drivers/usb/device/usb_dc_sam0.c]: They're much the same so my vote is to keep the enum.  SG?

**Review by eobalski — APPROVED** [review]

**Pull request merged**